### PR TITLE
[dynamo 3.11] support new jump opcodes

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94243
* #94102
* #94101
* #94100
* #94099
* #94098
* #93988
* #93987
* __->__ #93986

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/resume_execution.py [hunk 1]
```diff
@@ -163,6 +199,7 @@ def update(instructions: List[Instruction], code_options: Dict[str, Any]):
             assert not hooks
 
             prefix.append(create_instruction("JUMP_ABSOLUTE", target=target))
+            prefix.append(create_jump_absolute(target))
```
Comment (albanD) [inline torch/_dynamo/resume_execution.py:202]: double JUMP_ABSOLUTE here?

File: torch/_dynamo/bytecode_transformation.py [hunk 2]
```diff
@@ -57,6 +57,12 @@ def create_instruction(name, arg=None, argval=_NotProvided, target=None):
     )
 
 
+# Python 3.11 remaps
+def create_jump_absolute(target):
+    inst = "JUMP_FORWARD" if sys.version_info >= (3, 11) else "JUMP_ABSOLUTE"
+    return create_instruction(inst, target=target)
```
Comment (albanD) [inline torch/_dynamo/bytecode_transformation.py:63]: I'm surprised the target is the same for the both versions while one is an absolute value while the other is supposed to be a relative offset?

Comment (jansel) [reply torch/_dynamo/bytecode_transformation.py:63]: Inductor has a python byte code assembler lower down that converts object pointers into offsets.  We should make sure it knows how to generate JUMP_BACKWARD though.

**Review by jansel — CHANGES_REQUESTED** [review]
I think you need to update this function:
https://github.com/pytorch/pytorch/blob/f84f89b1c3f2bc74512e7a7b05ae6185164a9b3e/torch/_dynamo/bytecode_transformation.py#L174-L180

If the jump offset is negative, we should swap JUMP_FORWARD into JUMP_BACKWARD.

Please add a test for this.

**Comment by williamwen42** [comment]
@jansel do you want `devirtualize_jumps` need to be changed because you're concerned about negative values for oparg, or for optimization reasons? I looked into the implementation of jumps in cpython and it seems to support negative oparg jumps.

**Comment by jansel** [comment]
I was worried about correctness.  Why does JUMP_BACKWARD exist then?

Also, I think that offset calculation will be incorrect for JUMP_BACKWARD.

**Review by jansel — CHANGES_REQUESTED** [review]
Can you add some tests for codegen of these new jump opcodes?

**Comment by williamwen42** [comment]
@jansel I got the offset calculations wrong after all. I added tests and they should now be correct.

File: test/dynamo/test_misc.py [hunk 3]
```diff
@@ -3765,6 +3765,100 @@ def fn(x, y):
...
+                ]
+            )
+            return bytecode_transformation.clean_and_assemble_instructions(
+                insts,
+                list(code_options.keys()),
+                code_options,
+            )
+
+        # format: jump_opname, target_idx, expected forward jump, expected return value
+        test_args = (
```
Comment (albanD) [inline test/dynamo/test_misc.py:3834]: From this test, do I understand correctly that JUMP_FORWARD and JUMP_BACKWARD can be used the same way and the instructor creation will transform it in the right JUMP_* depending on the target?

Comment (williamwen42) [reply test/dynamo/test_misc.py:3834]: Yes.

File: torch/_dynamo/bytecode_transformation.py [hunk 4]
```diff
@@ -178,9 +184,26 @@ def devirtualize_jumps(instructions):
...
+                            inst.opname = inst.opname.replace("FORWARD", "BACKWARD")
+                            inst.opcode = dis.opmap[inst.opname]
+                            assert inst.opcode in jumps
+                    elif inst.arg > 0:
+                        # backward jumps become forward
+                        if sys.version_info >= (3, 11) and "BACKWARD" in inst.opname:
+                            inst.opname = inst.opname.replace("BACKWARD", "FORWARD")
+                            inst.opcode = dis.opmap[inst.opname]
+                            assert inst.opcode in jumps
+                    inst.arg //= 2
```
Comment (albanD) [inline torch/_dynamo/bytecode_transformation.py:205]: This doesn't work on 3.11? Do we need to binary search here?

**Review by albanD — APPROVED** [review]
Ok! Sounds good then!

**Review by malfet — APPROVED** [review]

File: torch/_dynamo/bytecode_transformation.py [hunk 5]
```diff
@@ -149,6 +155,22 @@ def virtualize_jumps(instructions):
                     break
 
 
+_REL_JUMPS = set(dis.hasjrel)
+
+
+def flip_jump_direction(instruction):
+    if sys.version_info < (3, 11):
+        raise RuntimeError("Cannot flip jump direction in Python < 3.11")
```
Comment (voznesenskym) [inline torch/_dynamo/bytecode_transformation.py:163]: assert instead of if/raise?

File: torch/_dynamo/bytecode_transformation.py [hunk 6]
```diff
@@ -170,17 +192,27 @@ def devirtualize_jumps(instructions):
...
-            else:  # relative jump
-                if sys.version_info < (3, 10):
-                    inst.arg = target.offset - inst.offset - instruction_size(inst)
                 else:
-                    inst.arg = int(
-                        (target.offset - inst.offset - instruction_size(inst)) / 2
-                    )
+                    raise RuntimeError("Python 3.11+ should not haave absolute jumps")
+            else:  # relative jump
+                inst.arg = int(target.offset - inst.offset - instruction_size(inst))
```
Comment (voznesenskym) [inline torch/_dynamo/bytecode_transformation.py:202]: Add a comment for what this does

File: torch/_dynamo/bytecode_transformation.py [hunk 7]
```diff
@@ -375,7 +407,10 @@ def transform_code_object(code, transformations, safe=False):
     propagate_line_nums(instructions)
 
     transformations(instructions, code_options)
+    return clean_and_assemble_instructions(instructions, keys, code_options)[1]
+
 
+def clean_and_assemble_instructions(instructions, keys, code_options):
```
Comment (voznesenskym) [inline torch/_dynamo/bytecode_transformation.py:413]: Put a type signature on this, esp if you are doing a multipart return

File: torch/_dynamo/resume_execution.py [hunk 8]
```diff
@@ -82,6 +82,42 @@ def __call__(self, code_options, cleanup):
                 create_instruction("POP_TOP"),
             ]
 
+        else:
+            # NOTE: copying over for now since more changes are anticipated
```
Comment (voznesenskym) [inline torch/_dynamo/resume_execution.py:86]: rm comment it makes no sense

**Review by voznesenskym — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**